service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/GrantPetUsagePermissionStep.java

```diff
@@ -0,0 +1,63 @@
+package bio.terra.workspace.service.resource.controlled.flight.create.notebook;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.service.petserviceaccount.PetSaService;
+import bio.terra.workspace.service.petserviceaccount.model.UserWithPetSa;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.PetSaKeys;
+import com.google.api.services.iam.v1.model.Policy;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Step to grant a user permission to use their pet service account. Controlled Notebooks run as a
+ * pet service account, and only users with the 'iam.serviceAccounts.actAs' permission on the pet
+ * are allowed through the proxy. GCP automatically de-deduplicates IAM bindings so this step is
+ * idempotent, even across flights.
+ */
+public class GrantPetUsagePermissionStep implements Step {
+
+  private static final Logger logger = LoggerFactory.getLogger(GrantPetUsagePermissionStep.class);
+
+  private final UUID workspaceId;
+  private final UserWithPetSa userAndPet;
+  private final PetSaService petSaService;
+
+  public GrantPetUsagePermissionStep(
+      UUID workspaceId, UserWithPetSa userAndPet, PetSaService petSaService) {
+    this.workspaceId = workspaceId;
+    this.userAndPet = userAndPet;
+    this.petSaService = petSaService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    Policy modifiedPolicy =
+        petSaService.enablePetServiceAccountImpersonation(workspaceId, userAndPet);
+    // Store the eTag value of the modified policy in case this step needs to be undone.
+    FlightMap workingMap = context.getWorkingMap();
+    workingMap.put(PetSaKeys.MODIFIED_PET_SA_POLICY_ETAG, modifiedPolicy.getEtag());
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    FlightMap workingMap = context.getWorkingMap();
+    String expectedEtag = workingMap.get(PetSaKeys.MODIFIED_PET_SA_POLICY_ETAG, String.class);
+    if (expectedEtag == null) {
```

Comment:
Is there a way to close this hole?
Maybe to have a step ahead of this that reads the current etag and saves it in the map. Then query the current etag here to see if it changed or not?

It doesn't solve concurrent updating, but we have not worried concurrency at all so far.

Reply:
updated javadoc for this and `RevokePetUsagePermissionStep`